fix(frontend): zoom/bbox state race produces "bbox too large" 400 mid-zoom-in

## Summary

When the user zooms in from a wide view (e.g. CONUS at z≈4) toward a small area (e.g. San Jose at z≥6), `/api/observations` returns:

```json
{"error":"bbox too large","maxLngSpan":45,"maxLatSpan":25,"hint":"zoom out for aggregated view"}
```

…followed ~250ms later by a successful request with the correct viewport bbox. The transient 400 is logged to the browser console and pollutes error telemetry; for slow networks or rapid re-zooms it can be the request the UI actually surfaces.

## Reproduction

Live-driven via Playwright MCP against `bird-maps.com`, 1440×900, 2026-05-20:

1. Load `https://bird-maps.com` — map mounts at CONUS (z=4, bbox≈`-129.5,26.9,-67.6,50.7`).
2. From the DevTools console, drive: `__map.jumpTo({ center: [-121.89, 37.34], zoom: 7 })`.
3. Observe network requests to `api.bird-maps.com/api/observations`:

| # | bbox (URL-decoded) | zoom | status |
|---|---|---|---|
| pre-jump | `-129.5,26.9,-67.6,50.7` (CONUS, 61.9° × 23.7°) | 4 | 200 |
| **mid-jump** | **`-129.5,26.9,-67.6,50.7` (still CONUS)** | **7** | **400 \"bbox too large\"** |
| ~250ms after | `-125.8,35.8,-118.0,38.9` (San Jose, 7.7° × 3.1°) | 7 | 200 |

Console:

```
[ERROR] Failed to load resource: 400 https://api.bird-maps.com/api/observations?since=14d&bbox=-129.5..&zoom=7
[ERROR] API error 400: {\"error\":\"bbox too large\",...}
```

The failing bbox spans 61.9° lng × 23.7° lat — exactly the CONUS framing from the previous viewport. Server-side cap (`services/read-api/src/validate.ts:179-206`) rejects `lngSpan > 45 OR latSpan > 25` at `zoom ≥ 6`.

## Root cause

`frontend/src/App.tsx:182-210` (`onViewportChange`) desyncs the two halves of the viewport snapshot:

```ts
const onViewportChange = useCallback((bounds: LngLatBounds, zoom: number) => {
  setViewportBounds(bounds);
  setDebouncedZoom(zoom);            // ← fires synchronously (despite the \"debounced\" name)
  const next: [number, number, number, number] = [
    bounds.getWest(), bounds.getSouth(), bounds.getEast(), bounds.getNorth(),
  ];
  if (bboxDebounceRef.current !== null) clearTimeout(bboxDebounceRef.current);
  bboxDebounceRef.current = setTimeout(() => {
    setDebouncedBbox(prev => /* no-op guard */);  // ← 250ms trailing-edge debounce
  }, 250);
}, []);
```

`MapCanvas.tsx:750-754` emits `(bounds, zoom)` paired atomically on each MapLibre `idle` event:

```ts
map.on('idle', () => {
  const cb = onViewportChangeRef.current;
  if (!cb) return;
  cb(map.getBounds(), Math.floor(map.getZoom()));
});
```

App.tsx breaks that pairing: `setDebouncedZoom(zoom)` updates immediately, `setDebouncedBbox(next)` updates 250ms later. During the 250ms window, `useBirdData` re-runs with `{ bbox: STALE_previous_viewport, zoom: NEW_zoom }`. When the new zoom crosses the `≥ 6` boundary and the previous bbox was a CONUS-sized frame, the server-side cap fires.

The variable name `debouncedZoom` is misleading — only the bbox is debounced. `debouncedZoom` has exactly one consumer (`useBirdData` at App.tsx:114); no other call site needs an immediate zoom value.

## Why this matters

1. Every zoom-in across the `< 6 → ≥ 6` boundary fires a 400 in console — noise that masks real errors and pollutes Clarity session recordings.
2. On slow networks the transient 400 can be the response that lands first relative to the React render cycle, briefly surfacing an error state to the user before the corrected request returns 200.
3. Counts against any future SLO on 4xx rate from the frontend.

## Proposed fix

Move `setDebouncedZoom(zoom)` **inside** the existing 250ms `setTimeout` so both halves of the viewport snapshot become atomic again. The rate-limit intent (per the comment at lines 170–174: \"rapid pan→zoom→pan sequences only trigger a single fetch\") is preserved; the `useCallback` closure captures the latest `(bounds, zoom)` per call, and `clearTimeout` ensures only the trailing-edge values land.

```ts
// frontend/src/App.tsx — onViewportChange
const onViewportChange = useCallback((bounds: LngLatBounds, zoom: number) => {
  setViewportBounds(bounds);
  const next: [number, number, number, number] = [
    bounds.getWest(), bounds.getSouth(), bounds.getEast(), bounds.getNorth(),
  ];
  if (bboxDebounceRef.current !== null) clearTimeout(bboxDebounceRef.current);
  bboxDebounceRef.current = setTimeout(() => {
    setDebouncedZoom(zoom);            // ← moved inside the timeout
    setDebouncedBbox(prev => {
      const epsilon = 1e-4;
      if (
        Math.abs(prev[0] - next[0]) < epsilon &&
        Math.abs(prev[1] - next[1]) < epsilon &&
        Math.abs(prev[2] - next[2]) < epsilon &&
        Math.abs(prev[3] - next[3]) < epsilon
      ) return prev;
      return next;
    });
  }, 250);
}, []);
```

Both state setters fire inside the same task; React 18 automatic batching coalesces them into a single render, so `useBirdData` re-runs once with consistent `{bbox, zoom}`.

Rename `debouncedZoom` → keep the name (it now actually is debounced) OR rename the pair to `viewportSnapshot` if a small follow-up refactor is in scope. Out of scope for the fix itself.

## Test plan

1. **Unit (regression)** — `frontend/src/App.test.tsx`: drive `onViewportChange` with `(CONUS_bounds, 4)`, advance fake timers by 250ms, then drive `(SJ_bounds, 7)`, advance 250ms. Assert exactly one fetch per advance, each with consistent `{bbox, zoom}`. Failing case before fix: fetch fires immediately after the second call with `{bbox: CONUS, zoom: 7}` — the bad combination.
2. **E2E** — `frontend/e2e/`: navigate to `/`, wait for map idle, expose `__map`, `jumpTo({ center: [-121.89, 37.34], zoom: 7 })`, wait 500ms, assert no 400s in network log for `/api/observations`.
3. **Manual** — repeat the Playwright MCP repro above; the mid-jump 400 should not appear.

## Alternatives considered

- **Drop the debounce entirely.** Relies on MapLibre's `idle` being already once-per-settle. Rejected — the existing comment documents real protection against rapid sequential settles (`pan→zoom→pan` gesture chains); removing it would 3× the API rate during interactive use.
- **Server-side: relax the cap when zoom is monotonically increasing.** Stateful, dangerous, doesn't address the root pairing bug.
- **Move bbox + zoom into a single `useState({bbox, zoom})` object.** Cleanest long-term, larger blast radius (touches `useBirdData` filter shape). Defer to a follow-up.

## Evidence artifacts

- Frontend code: `frontend/src/App.tsx:182-210`, `frontend/src/components/map/MapCanvas.tsx:750-754`
- Server cap: `services/read-api/src/validate.ts:179-206`
- Reproduction date: 2026-05-20
- Reproduction tool: Playwright MCP against production `bird-maps.com`

--- a/frontend/src/App.test.tsx
+++ b/frontend/src/App.test.tsx
@@ -1,6 +1,7 @@
 import { describe, it, expect, vi, afterEach, beforeEach } from 'vitest';
 import { render, screen, waitFor, act } from '@testing-library/react';
 import userEvent from '@testing-library/user-event';
+import type { LngLatBounds } from 'maplibre-gl';
 
 // Phase 4 / #663: useIsCompact calls window.matchMedia. JSDOM does not
 // implement it — polyfill with a stub that returns non-compact (wide
@@ -18,7 +19,13 @@ vi.stubGlobal('matchMedia', (query: string) => ({
 }));
 
 // Hoist mock fns so they exist before any module-level code runs.
-const { mockGetHotspots, mockGetObservations, mockGetSilhouettes, mockUrlState } = vi.hoisted(() => ({
+const {
+  mockGetHotspots,
+  mockGetObservations,
+  mockGetSilhouettes,
+  mockUrlState,
+  mapSurfaceRef,
+} = vi.hoisted(() => ({
   mockGetHotspots: vi.fn(),
   mockGetObservations: vi.fn(),
   mockGetSilhouettes: vi.fn(),
@@ -32,6 +39,15 @@ const { mockGetHotspots, mockGetObservations, mockGetSilhouettes, mockUrlState }
     },
     set: vi.fn(),
   },
+  // Capture handle for the zoom/bbox state-race regression (issue #690).
+  // The MapSurface stub assigns the latest `onViewportChange` prop here on
+  // every render so the test can drive App.tsx's viewport callback the same
+  // way MapCanvas's `idle` event would.
+  mapSurfaceRef: {
+    onViewportChange: null as
+      | ((bounds: unknown, zoom: number) => void)
+      | null,
+  },
 }));
 
 // Stub url-state before App imports it.
@@ -44,10 +60,20 @@ vi.mock('./state/url-state.js', () => ({
 // color-SOT wiring added a silhouettes fetch that runs concurrently with
 // App's initial render, the 'map' view aria-busy test is now reliably hot
 // enough to trigger MapSurface's maplibre import — which fails in jsdom
-// because `window.URL.createObjectURL` isn't polyfilled. The test only
-// cares about the `<main aria-busy>` attribute, not the map itself.
+// because `window.URL.createObjectURL` isn't polyfilled. Most tests in this
+// file only care about the `<main aria-busy>` attribute, not the map itself.
+//
+// Issue #690: the stub additionally captures the latest `onViewportChange`
+// prop into `mapSurfaceRef.onViewportChange` so the zoom/bbox state-race
+// regression test can invoke it directly. Mirrors the `MapSurface.test.tsx`
+// pattern at lines 18–29.
 vi.mock('./components/MapSurface.js', () => ({
-  MapSurface: () => null,
+  MapSurface: (props: {
+    onViewportChange?: (bounds: unknown, zoom: number) => void;
+  }) => {
+    mapSurfaceRef.onViewportChange = props.onViewportChange ?? null;
+    return null;
+  },
 }));
 
 // Stub the ApiClient constructor so useBirdData receives a controllable mock.
@@ -678,5 +704,114 @@ describe('Clarity view tagging (#657-followup)', () => {
     };
     render(<App />);
     expect(setViewSpy).toHaveBeenCalledWith('feed');
+  });
+});
+
+describe('Zoom/bbox state-race regression (#690)', () => {
+  // Build a duck-typed LngLatBounds for the onViewportChange callback. App.tsx
+  // only ever calls .getWest/.getSouth/.getEast/.getNorth on the value, so a
+  // plain object with those methods is sufficient and lets us avoid pulling in
+  // maplibre-gl in jsdom (the same constraint that drives the MapSurface stub
+  // above).
+  function makeBounds(
+    west: number, south: number, east: number, north: number,
+  ): LngLatBounds {
+    return {
+      getWest: () => west,
+      getSouth: () => south,
+      getEast: () => east,
+      getNorth: () => north,
+    } as unknown as LngLatBounds;
+  }
+
+  // CONUS framing (≈61.9° lng × 23.7° lat) — the bbox MapCanvas reports on
+  // the initial `idle` from the default zoom-4 view. Span exceeds the server's
+  // 45° lng / 25° lat cap, so any /api/observations call carrying this bbox
+  // with zoom ≥ 6 is the bad combination this regression guards against.
+  const CONUS_BOUNDS = makeBounds(-125, 24, -66, 50);
+  // San Jose framing (≈7.8° lng × 3.1° lat) — well inside the cap at any zoom.
+  const SJ_BOUNDS = makeBounds(-125.8, 35.8, -118.0, 38.9);
+
+  beforeEach(() => {
+    __resetSilhouettesCache();
+    mockGetHotspots.mockResolvedValue([]);
+    mockGetObservations.mockResolvedValue({
+      data: [], meta: { freshestObservationAt: null },
+    });
+    mockGetSilhouettes.mockResolvedValue([]);
+    mapSurfaceRef.onViewportChange = null;
+    mockUrlState.state = {
+      since: '14d', notable: false, speciesCode: null, familyCode: null,
+      view: 'map',
+    };
+  });
+
+  afterEach(() => {
+    vi.useRealTimers();
+    vi.restoreAllMocks();
+  });
+
+  it('never emits a {bbox: CONUS, zoom: ≥6} fetch when zooming from CONUS into San Jose', async () => {
+    vi.useFakeTimers({ shouldAdvanceTime: true });
+
+    render(<App />);
+
+    // Wait for App to finish mounting and render <MapSurface>. The stub
+    // captures onViewportChange on each render; we poll until the handle is
+    // populated so the rest of the test can drive viewport changes.
+    await waitFor(() => {
+      expect(mapSurfaceRef.onViewportChange).not.toBeNull();
+    });
+    const onViewportChange = mapSurfaceRef.onViewportChange!;
+
+    // Step 1: simulate the initial CONUS settle at zoom 4. App schedules the
+    // 250ms bbox debounce; setDebouncedZoom fires immediately.
+    await act(async () => {
+      onViewportChange(CONUS_BOUNDS, 4);
+    });
+    await act(async () => {
+      await vi.advanceTimersByTimeAsync(250);
+    });
+
+    // Step 2: simulate jumpTo({center: SJ, zoom: 7}) — MapCanvas emits a
+    // single onViewportChange(SJ, 7) on the next idle.
+    await act(async () => {
+      onViewportChange(SJ_BOUNDS, 7);
+    });
+    // The bug — if present — fires a synchronous re-render with
+    // {bbox: CONUS, zoom: 7} BEFORE the 250ms timeout drains. Drain any
+    // microtasks so useBirdData's effect runs against the current state.
+    await act(async () => {
+      await vi.advanceTimersByTimeAsync(0);
+    });
+
+    // Step 3: drain the bbox debounce — App now commits {bbox: SJ, zoom: 7},
+    // which is a consistent pairing under the cap.
+    await act(async () => {
+      await vi.advanceTimersByTimeAsync(250);
+    });
+
+    // Assert every fetch carries a consistent {bbox, zoom} pair: either the
+    // span fits the server's 45° lng cap, or the zoom is below the gating
+    // threshold (cap at services/read-api/src/validate.ts only fires when
+    // zoom ≥ 6). The pre-fix race produces a call with lngSpan ≈ 59 AND
+    // zoom = 7 — violating both halves of the predicate.
+    const calls = mockGetObservations.mock.calls;
+    expect(calls.length).toBeGreaterThan(0);
+    for (const [filters] of calls) {
+      const { bbox, zoom } = filters as {
+        bbox?: [number, number, number, number];
+        zoom?: number;
+      };
+      // The initial mount call may omit bbox/zoom; the assertion only applies
+      // once a viewport-derived pair has been threaded through.
+      if (!bbox || zoom === undefined) continue;
+      const lngSpan = bbox[2] - bbox[0];
+      const consistent = lngSpan <= 45 || zoom < 6;
+      expect(
+        consistent,
+        `inconsistent fetch: bbox=${bbox.join(',')} (lngSpan=${lngSpan.toFixed(2)}), zoom=${zoom}`,
+      ).toBe(true);
+    }
   });
 });

--- a/frontend/src/App.tsx
+++ b/frontend/src/App.tsx
@@ -181,7 +181,6 @@ export function App() {
   );
   const onViewportChange = useCallback((bounds: LngLatBounds, zoom: number) => {
     setViewportBounds(bounds);
-    setDebouncedZoom(zoom);
     const next: [number, number, number, number] = [
       bounds.getWest(),
       bounds.getSouth(),
@@ -190,6 +189,17 @@ export function App() {
     ];
     if (bboxDebounceRef.current !== null) clearTimeout(bboxDebounceRef.current);
     bboxDebounceRef.current = setTimeout(() => {
+      // Issue #690: setDebouncedZoom MUST fire inside the same task as
+      // setDebouncedBbox so {bbox, zoom} stays paired in the useBirdData
+      // filters. Firing the zoom setter synchronously above (outside the
+      // timeout) decoupled the pair — during the 250ms window the effect
+      // would re-run with {bbox: STALE_previous_viewport, zoom: NEW_zoom},
+      // which trips the server's bbox-area cap (services/read-api/src/
+      // validate.ts assertBboxAreaCap: zoom ≥ 6 && lngSpan > 45 → 400
+      // "bbox too large") on any zoom-in across the < 6 → ≥ 6 boundary.
+      // React 18 automatic batching coalesces both setters into a single
+      // render so useBirdData re-fires exactly once with a consistent pair.
+      setDebouncedZoom(zoom);
       setDebouncedBbox(prev => {
         // No-op guard: skip the state update (and the consequent refetch)
         // if the bbox hasn't moved meaningfully. ~1e-4 degrees ≈ 10m at

